File: spicedb-go/client/raw_escape_hatch_test.go
package client

import (
	"context"
	"net"
	"reflect"
	"testing"

	"github.com/stretchr/testify/require"
	"google.golang.org/grpc"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/test/bufconn"

	v1 "github.com/authzed/spicedb-clients/proto-clients/spicedb-go-proto/gen/authzed/api/v1"
	"github.com/authzed/spicedb-clients/spicedb-go/consistency"
	"github.com/authzed/spicedb-clients/spicedb-go/rel"
)

// The escape hatch, RawProto, exists so a request the idiomatic surface cannot
// express has a workaround short of forking the client. Asserting the accessor
// returns something non-nil would prove none of that. What matters is whether a
// caller can drive a generated service client through it and get an answer out
// of a real server, with this client's bearer token attached -- so the tests
// below run a real (in-process, bufconn-backed) gRPC server and assert the
// authorization metadata the handler actually received.
//
// The RPC driven here is CheckPermission, the single-check call the idiomatic
// client never makes (CheckOne routes every check through
// CheckBulkPermissions), so the gap is genuine rather than contrived.

// authRecordingServer implements both the single-check and bulk-check RPCs,
// recording the authorization metadata of every request in arrival order.
type authRecordingServer struct {
	v1.UnimplementedPermissionsServiceServer

	authorizations []string
}

func (s *authRecordingServer) record(ctx context.Context) {
	md, _ := metadata.FromIncomingContext(ctx)
	values := md.Get("authorization")
	if len(values) == 0 {
		s.authorizations = append(s.authorizations, "")
		return
	}
	s.authorizations = append(s.authorizations, values[0])
}

func (s *authRecordingServer) CheckPermission(ctx context.Context, _ *v1.CheckPermissionRequest) (*v1.CheckPermissionResponse, error) {
	s.record(ctx)
	return &v1.CheckPermissionResponse{
		Permissionship: v1.CheckPermissionResponse_PERMISSIONSHIP_HAS_PERMISSION,
		CheckedAt:      &v1.ZedToken{Token: "rev-raw"},
	}, nil
}

func (s *authRecordingServer) CheckBulkPermissions(ctx context.Context, req *v1.CheckBulkPermissionsRequest) (*v1.CheckBulkPermissionsResponse, error) {
	s.record(ctx)
	pairs := make([]*v1.CheckBulkPermissionsPair, len(req.GetItems()))
	for i := range req.GetItems() {
		pairs[i] = &v1.CheckBulkPermissionsPair{
			Response: &v1.CheckBulkPermissionsPair_Item{
				Item: &v1.CheckBulkPermissionsResponseItem{
					Permissionship: v1.CheckPermissionResponse_PERMISSIONSHIP_HAS_PERMISSION,
				},
			},
		}
	}
	return &v1.CheckBulkPermissionsResponse{Pairs: pairs}, nil
}

func newAuthRecordingTestClient(t *testing.T) (*Client, *authRecordingServer) {
	t.Helper()

	const bufSize = 1024 * 1024
	lis := bufconn.Listen(bufSize)

	srv := &authRecordingServer{}
	grpcServer := grpc.NewServer()
	v1.RegisterPermissionsServiceServer(grpcServer, srv)
	go func() { _ = grpcServer.Serve(lis) }()
	t.Cleanup(grpcServer.Stop)

	c, err := NewWithOpts("passthrough:///bufnet", "test-token",
		WithInsecure(),
		// bufnet is an in-memory bufconn dial target, not a real network
		// destination -- unrelated to what these tests exercise.
		WithInsecureAllowRemoteHost(),
		WithDialOptions(grpc.WithContextDialer(func(ctx context.Context, _ string) (net.Conn, error) {
			return lis.DialContext(ctx)
		})),
	)
	require.NoError(t, err)
	t.Cleanup(func() { _ = c.Close() })
	return c, srv
}

func checkPermissionRequest() *v1.CheckPermissionRequest {
	return &v1.CheckPermissionRequest{
		Consistency: &v1.Consistency{
			Requirement: &v1.Consistency_FullyConsistent{FullyConsistent: true},
		},
		Resource:   &v1.ObjectReference{ObjectType: "document", ObjectId: "readme"},
		Permission: "view",
		Subject: &v1.SubjectReference{
			Object: &v1.ObjectReference{ObjectType: "user", ObjectId: "jimmy"},
		},
	}
}

func TestRawProto_DrivesARealServiceClientAgainstARealServer(t *testing.T) {
	c, srv := newAuthRecordingTestClient(t)

	resp, err := c.RawProto().PermissionsServiceClient.CheckPermission(context.Background(), checkPermissionRequest())
	require.NoError(t, err)
	require.Equal(t, v1.CheckPermissionResponse_PERMISSIONSHIP_HAS_PERMISSION, resp.GetPermissionship())
	require.Equal(t, "rev-raw", resp.GetCheckedAt().GetToken())

	// The bearer token rides this client's own connection credentials, so a
	// raw call is authenticated exactly as an idiomatic one is.
	require.Equal(t, []string{"Bearer test-token"}, srv.authorizations)
}

func TestRawProto_SharesTheConnectionTheIdiomaticMethodsUse(t *testing.T) {
	c, srv := newAuthRecordingTestClient(t)

	// Not a second connection built behind the caller's back.
	require.Same(t, c.RawProto(), c.RawProto())

	result, err := c.CheckOne(context.Background(), consistency.Full(), "view",
		rel.MustFromTriple("document", "readme", "view", "user", "jimmy", ""))
	require.NoError(t, err)
	require.True(t, result.HasPermission())

	_, err = c.RawProto().PermissionsServiceClient.CheckPermission(context.Background(), checkPermissionRequest())
	require.NoError(t, err)

	// One idiomatic call (via CheckBulkPermissions) and one raw call (via the
	// single-check RPC), both authenticated, both on this client's own
	// connection.
	require.Equal(t, []string{"Bearer test-token", "Bearer test-token"}, srv.authorizations)
}

// The hatch must never grow into a way to build a connection. Root DESIGN.md,
// "RULE: Credentials over insecure transport require an explicit opt-in", is
// enforced in NewWithOpts, on the single path that dials. Handing back an
// already-built client cannot bypass that; accepting an endpoint, token, or
// transport setting would.
func TestRawProto_IsAnAccessorNotASecondConstructionPath(t *testing.T) {
	method, ok := reflect.TypeOf(&Client{}).MethodByName("RawProto")
	require.True(t, ok, "RawProto must be exported")
	// One parameter: the receiver. Anything more is a construction argument.
	require.Equal(t, 1, method.Type.NumIn(), "RawProto must take no arguments")

	// And the guard still refuses what it always did.
	_, err := NewWithOpts("evil.example.com:50051", "test-token", WithInsecure())
	require.Error(t, err)
	require.Contains(t, err.Error(), "WithInsecureAllowRemoteHost")
}

func TestRawProto_OnAZeroValueClientIsNil(t *testing.T) {
	require.Nil(t, (&Client{}).RawProto())
	var nilClient *Client
	require.Nil(t, nilClient.RawProto())
}

File: spicedb-typescript/src/__tests__/raw-escape-hatch.test.ts
/**
 * The escape hatch reaches a real stub and makes a real call.
 *
 * `raw()` exists so a request the idiomatic surface cannot express has a
 * workaround short of forking the client -- root DESIGN.md, "What NOT To Do",
 * permits exactly this as "clearly marked secondary API". Asserting that the
 * accessor exists, or that it returns something with a `permissions` property,
 * would prove none of that. What matters is whether a caller can drive a
 * generated Connect client through it and get an answer out of a real server,
 * with this client's bearer token attached.
 *
 * So these tests run a real plaintext HTTP/2 (h2c) server behind
 * `connectNodeAdapter`, and check both the response and the `authorization`
 * header the server actually received. No TLS, so unlike custom-tls.test.ts
 * these need no `openssl`.
 */
import { describe, it, expect } from "vitest";
import * as http2 from "node:http2";
import { connectNodeAdapter } from "@connectrpc/connect-node";
import { PermissionsService } from "@spicedb/proto";
import { SpiceDBClient } from "../client.js";
import { full } from "../consistency.js";

const TOKEN = "test-token";

/**
 * Serves `CheckBulkPermissions` for real over h2c, recording the
 * `authorization` header of every request that arrives.
 */
async function serve(): Promise<{
  port: number;
  authorizations: string[];
  close: () => Promise<void>;
}> {
  const authorizations: string[] = [];
  const handler = connectNodeAdapter({
    routes: (router) => {
      router.service(PermissionsService, {
        checkBulkPermissions: (req, ctx) => {
          authorizations.push(ctx.requestHeader.get("authorization") ?? "");
          return {
            pairs: req.items.map(() => ({
              response: {
                case: "item" as const,
                value: { permissionship: 2 }, // HAS_PERMISSION
              },
            })),
          };
        },
      });
    },
  });
  const server = http2.createServer(handler);
  await new Promise<void>((resolve) => server.listen(0, "localhost", resolve));
  const address = server.address();
  if (address === null || typeof address === "string") {
    throw new Error("expected a TCP address from the test server");
  }
  return {
    port: address.port,
    authorizations,
    close: () => new Promise<void>((resolve) => server.close(() => resolve())),
  };
}

function bulkRequest() {
  return {
    consistency: {
      requirement: { case: "fullyConsistent" as const, value: true },
    },
    items: [
      {
        resource: { objectType: "document", objectId: "readme" },
        permission: "view",
        subject: { object: { objectType: "user", objectId: "jimmy" } },
      },
    ],
  };
}

describe("SpiceDBClient raw()", () => {
  it("drives a real generated client against a real server", async () => {
    const { port, authorizations, close } = await serve();
    const client = new SpiceDBClient({
      endpoint: `localhost:${port}`,
      token: TOKEN,
      insecure: true,
      maxRetries: 0,
    });
    try {
      const response =
        await client.raw().permissions.checkBulkPermissions(bulkRequest());

      expect(response.pairs).toHaveLength(1);
      expect(response.pairs[0].response.case).toBe("item");
      // The bearer token rides the transport interceptor, so a raw call is
      // authenticated exactly as an idiomatic one is.
      expect(authorizations).toEqual([`Bearer ${TOKEN}`]);
    } finally {
      client.close();
      await close();
    }
  }, 30_000);

  it("hands back the very client the idiomatic methods call through", async () => {
    const { port, authorizations, close } = await serve();
    const client = new SpiceDBClient({
      endpoint: `localhost:${port}`,
      token: TOKEN,
      insecure: true,
      maxRetries: 0,
    });
    try {
      // Not a second connection built behind the caller's back: the same
      // object every time, and the same one an idiomatic call uses.
      expect(client.raw()).toBe(client.raw());

      const idiomatic = await client.checkPermissions(full(), {
        resourceType: "document",
        resourceId: "readme",
        permission: "view",
        subjectType: "user",
        subjectId: "jimmy",
      });
      expect(idiomatic).toHaveLength(1);

      await client.raw().permissions.checkBulkPermissions(bulkRequest());

      // One idiomatic call and one raw call, both authenticated, both over the
      // connection this client owns.
      expect(authorizations).toEqual([`Bearer ${TOKEN}`, `Bearer ${TOKEN}`]);
    } finally {
      client.close();
      await close();
    }
  }, 30_000);

  it("is an accessor, not a second construction path", () => {
    // Root DESIGN.md, "RULE: Credentials over insecure transport require an
    // explicit opt-in", is enforced in the proto tier's createSpiceDBClient,
    // on the single path that builds a transport. Handing back an
    // already-built client cannot bypass that; accepting an endpoint, token,
    // or transport setting here would -- it would be a second construction
    // path with no guard on it. Pin the shape that makes that impossible.
    expect(SpiceDBClient.prototype.raw.length).toBe(0);

    // And the guard still refuses the combination it always did, so the hatch
    // has not moved construction anywhere.
    expect(
      () =>
        new SpiceDBClient({
          endpoint: "evil.example.com:50051",
          token: TOKEN,
          insecure: true,
        }),
    ).toThrow(/allowInsecureRemoteCredentials/);
  });
});
